Simplenote/src/main/java/com/automattic/simplenote/SignInActivity.java

```diff
@@ -0,0 +1,184 @@
+
+    private String mAuthState;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        // Ensure we are working with the expected layout from Simperium's `LoginActivity`.
+        View mainView = findViewById(R.id.main);
+        if (!(mainView instanceof ScrollView)) {
+            return;
+        }
+
+        ScrollView scrollView = (ScrollView)mainView;
+        if (scrollView.getChildAt(0) == null || !(scrollView.getChildAt(0) instanceof LinearLayout)) {
+            return;
+        }
+
+        // Create and add a new view that contains the wp.com sign in button
+        LinearLayout layout = (LinearLayout)scrollView.getChildAt(0);
+        View signInFooter = getLayoutInflater().inflate(R.layout.sign_in_footer, layout, false);
+
+        View footerButton = signInFooter.findViewById(R.id.wpcom_signin_button);
+        footerButton.setOnClickListener(mWPSignInButtonClickListener);
+
+        layout.addView(signInFooter);
+    }
+
+    private View.OnClickListener mWPSignInButtonClickListener = new View.OnClickListener() {
+        @Override
+        public void onClick(View v) {
+            AuthorizationServiceConfiguration serviceConfig = new AuthorizationServiceConfiguration(
+                    Uri.parse(WPCOM_OAUTH_URL + "authorize"),
+                    Uri.parse(WPCOM_OAUTH_URL + "token"));
+
+            Uri redirectUri = Uri.parse(WPCOM_OAUTH_REDIRECT_URL);
+            AuthorizationRequest.Builder authRequestBuilder =
+                    new AuthorizationRequest.Builder(
+                            serviceConfig,
+                            BuildConfig.WPCOM_CLIENT_ID,
+                            ResponseTypeValues.CODE,
+                            redirectUri);
+
+            // Set a unique state value
+            mAuthState = "app-" + UUID.randomUUID();
+            authRequestBuilder.setState(mAuthState);
+
+            AuthorizationRequest request = authRequestBuilder.build();
+            AuthorizationService authService = new AuthorizationService(SignInActivity.this);
+            Intent authIntent = authService.getAuthorizationRequestIntent(request);
+            startActivityForResult(authIntent, OAUTH_ACTIVITY_CODE);
+        }
+    };
+
+    @Override
+    protected void onSaveInstanceState(Bundle outState) {
+        outState.putString(STATE_KEY_AUTH, mAuthState);
+
+        super.onSaveInstanceState(outState);
+    }
+
+    @Override
+    protected void onRestoreInstanceState(Bundle savedInstanceState) {
+        super.onRestoreInstanceState(savedInstanceState);
+
+        if (savedInstanceState.containsKey(STATE_KEY_AUTH)) {
+            mAuthState = savedInstanceState.getString(STATE_KEY_AUTH);
+        }
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == OAUTH_ACTIVITY_CODE) {
```

Comment:
holy nesting batman. what if we performed an early-abort? this `if` doesn't even appear to have an `else`…

```java
if (requestCode != OAUTH_ACTIVITY_CODE) {
    return;
}

AuthorizationResponse authResponse…
```

and same going through this deeply-nested conditional tree

Reply:
I think I'll flatten it, I did similar early returns in other areas so best to match it up.